WalletWasabi.Tests/UnitTests/WabiSabi/Client/AmountDecomposerTests.cs

```diff
@@ -28,24 +28,46 @@ public class AmountDecomposerTests
 	[InlineData(5000, 0, 8)]
-	public void DecompositionsInvariantTest(decimal feeRateDecimal, long minOutputAmount, int maxAvailableOutputs)
+	[InlineData(0, 0, 8, true)]
+	[InlineData(0, 0, 1, true)]
+	[InlineData(0, 0, 2, true)]
+	[InlineData(0, 0, 3, true)]
+	[InlineData(0, 1_000, 1, true)]
+	[InlineData(0, 100_000, 2, true)]
+	[InlineData(0, 1_000_000, 3, true)]
+	[InlineData(0, 10_000_000, 8, true)]
+	[InlineData(20, 0, 1, true)]
+	[InlineData(100, 0, 2, true)]
+	[InlineData(500, 0, 3, true)]
+	[InlineData(5000, 0, 8, true)]
+	public void DecompositionsInvariantTest(decimal feeRateDecimal, long minOutputAmount, int maxAvailableOutputs, bool isTaprootEnabled = false)
 	{
-		var availableVsize = maxAvailableOutputs * Constants.P2wpkhOutputVirtualSize;
+		var outputVirtualSize = isTaprootEnabled ? Constants.P2trOutputVirtualSize : Constants.P2wpkhOutputVirtualSize;
+		var availableVsize = maxAvailableOutputs * outputVirtualSize;
 		var feeRate = new FeeRate(feeRateDecimal);
-		var feePerOutput = feeRate.GetFee(Constants.P2wpkhOutputVirtualSize);
+		var feePerOutput = feeRate.GetFee(outputVirtualSize);
 		var registeredCoinEffectiveValues = GenerateRandomCoins().Take(3).Select(c => c.EffectiveValue(feeRate, CoordinationFeeRate.Zero)).ToList();
 		var theirCoinEffectiveValues = GenerateRandomCoins().Take(30).Select(c => c.EffectiveValue(feeRate, CoordinationFeeRate.Zero)).ToList();
 		var allowedOutputAmountRange = new MoneyRange(Money.Satoshis(minOutputAmount), Money.Satoshis(ProtocolConstants.MaxAmountPerAlice));
 
-		var amountDecomposer = new AmountDecomposer(feeRate, allowedOutputAmountRange, Constants.P2wpkhOutputVirtualSize, Constants.P2wpkhInputVirtualSize, availableVsize);
+		var amountDecomposer = new AmountDecomposer(feeRate, allowedOutputAmountRange, availableVsize, isTaprootEnabled, Random);
 		var outputValues = amountDecomposer.Decompose(registeredCoinEffectiveValues, theirCoinEffectiveValues);
 
 		var totalEffectiveValue = registeredCoinEffectiveValues.Sum(x => x);
-		var totalEffectiveCost = outputValues.Count() * feePerOutput;
+		var totalEffectiveCost = outputValues.Sum(x => x.EffectiveCost);
 
-		Assert.InRange(outputValues.Count(), 1, maxAvailableOutputs);
-		Assert.True(totalEffectiveValue - totalEffectiveCost - minOutputAmount <= outputValues.Sum());
-		Assert.All(outputValues, v => Assert.InRange(v.Satoshi, minOutputAmount, totalEffectiveValue));
+		if (!isTaprootEnabled)
+		{
+			Assert.InRange(outputValues.Count(), 1, maxAvailableOutputs);
+		}
+		else
+		{
+			// The number of outputs cannot be ensure bacause of random scriptype generation. Instead we verify the total.
+			Assert.InRange(outputValues.Sum(x => x.ScriptType.EstimateOutputVsize()), 1, availableVsize);
+		}
+
+		Assert.True(totalEffectiveValue - totalEffectiveCost - minOutputAmount <= outputValues.Sum(x => x.EffectiveCost));
+		Assert.All(outputValues, v => Assert.InRange(v.EffectiveCost.Satoshi, minOutputAmount, totalEffectiveValue));
```

Comment:
I think the test should assert than when taproot is not enabled we don't have taproot outputs at least. It would be also good to assert that when taproot is enabled we have taproot outputs but that cannot be done in this test, what means we would need a new test.

Reply:
> when taproot is not enabled we don't have taproot outputs at least

Added a check.

> when taproot is enabled we have taproot outputs

It is not guaranteed - do we have a requirement like that?

Reply:
> It is not guaranteed

It is by probability. If the probability of getting 1 taproot output is 0.5 then after generating 100 outputs the probability of having 0 taproot outputs is $2^{-100}$, that means 0 in practical terms.

> do we have a requirement like that?

Well allowing taproot outputs is the reason of this PR.

Reply:
The decomposer receives a random generator so, it is probabilistic but also deterministic.